[REEF-1211] multi-runtime implementation

shulmanb opened the pull request:
This changeset addresses the issue by:
  1)Providing multi-runtime implementation
  2)Unit tests for multi runtime implementation
  3)Yarn test for Yarn/Multi runtime
  4)HelloREEF example for Yarn/Local multi runtime

JIRA:
  [REEF-1211](https://issues.apache.org/jira/browse/REEF-1211)

shulmanb commented:
@afchung, @markusweimer
Please take a look.

markusweimer commented:
This needs some more description on the JIRA. I left a note there.

markusweimer commented:
@shulmanb Is this "done" I see that there have been commits since you posted it and asked for review?

afchung commented:
@shulmanb Is it possible to split this into more manageable chunks? This PR seems to cover a lot of items.

shulmanb commented:
@afchung I prefer not to split it as it is hard to split it to parts that compile separatly

shulmanb commented:
@markusweimer  this is done. Fixed some build errors after the initial commit

shulmanb commented:
@afchung Most of the logic is in one class RuntimesHost and the event handlers just delegate calls to it. Also I prefer having tests and code in the same checkin

markusweimer commented on lang/java/reef-examples/pom.xml:
```diff
@@ -102,6 +102,11 @@ under the License.
+        <dependency>
+            <groupId>org.apache.reef</groupId>
+            <artifactId>reef-runtime-multi</artifactId>
+            <version>0.14.0-SNAPSHOT</version>
```
Replace with `${project.version}`

markusweimer commented on lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloMultiDriver.java:
```diff
@@ -0,0 +1,95 @@
+   */
+  public final class EvaluatorAllocatedHandler implements EventHandler<AllocatedEvaluator> {
+    @Override
+    public void onNext(final AllocatedEvaluator allocatedEvaluator) {
```
It would be good to separate different tasks to different runtimes here to highlight this ability.

shulmanb replied:
I need to create a better example, probably tasks need to send different messages back, as it is hard to see output for the local runtime evaluator. I will do that post this checkin.

markusweimer replied:
Please file a JIRA for that such that we won't forget.

markusweimer replied:
Did the JIRA get filled?

shulmanb replied:
Now it got filed: https://issues.apache.org/jira/browse/REEF-1230

markusweimer replied:
Thanks! I left some comments over there.

markusweimer commented on lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloREEFMultiYarn.java:
```diff
@@ -0,0 +1,95 @@
+        .build();
+  }
+
+  private static Configuration getHybridYarnSubmissionRuntimeConfiguration() {
```
This looks wonky. Where does the need for a `HashMap` come from?

shulmanb replied:
I need to bind named parameters for the original runtimes (local runtime in this case). I cannot bind them outside as tehy do not belong to that module. SO I need to pass them as values and bind them internally. Couldn't find any other way to make it work.

markusweimer commented on lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloTaskDelayed.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * A 'hello REEF' Task.
+ */
+public final class HelloTaskDelayed implements Task {
```
Why do we need this delayed task? Can't we reuse the other one? A 5 sec wait on these tasks will kill us as we add this to the integration tests.

shulmanb replied:
It looks like there is a bug in local runtime. Looks like a race condition between evaluator communication creation and task completion, so without the delay the task actually exits before communication channel is established and we cannot send notification back. This results in unclear shutdown, just wanted to see clear logs for this runtime. I agree that we need to fix this but don't want to wait for this. Will create JIRA for this issue.

markusweimer replied:
This is odd. The same code runs just fine on the local runtime without the wait time.

markusweimer commented on lang/java/reef-runtime-multi/pom.xml:
```diff
@@ -100,6 +122,14 @@ under the License.
+        <dependency>
+            <groupId>org.apache.reef</groupId>
+            <artifactId>reef-runtime-hdinsight</artifactId>
+            <version>0.14.0-SNAPSHOT</version>
```
Replace with `{project.version}`

markusweimer commented on lang/java/reef-runtime-multi/src/main/avro/RuntimeDefinition.avsc:
```diff
@@ -0,0 +1,30 @@
+ */
+ [
+{
+    "namespace":"org.apache.reef.runtime.multi.utils",
```
Please add documentation to this schema.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/AbstractDriverConfigurationProvider.java:
```diff
@@ -0,0 +1,117 @@
+/**
+ * Provides base class for driver configuration providers for multi runtimes.
+ */
+public abstract class AbstractDriverConfigurationProvider implements DriverConfigurationProvider {
```
Does this have to be `public`? Also, I generally am concerned about `abstract` classes due to the fragile base class problem. Any chance to do this via composition instead?

shulmanb replied:
It doesn't need to be public. I actually always prefer inheritance over composition, but it is a matter of taste.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeHDILocalDriverConfigurationProviderImpl.java:
```diff
@@ -0,0 +1,83 @@
+/**
+ * MultiRuntime configuration provider for HDI/Local runtime.
+ */
+public class MultiRuntimeHDILocalDriverConfigurationProviderImpl extends AbstractDriverConfigurationProvider {
```
Does this have to be `public`?

shulmanb replied:
Yes as we need to bind this in the actual app.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeHDILocalDriverConfigurationProviderImpl.java:
```diff
@@ -0,0 +1,83 @@
+  private final Set<String> rackNames;
+
+  @Inject
+  public MultiRuntimeHDILocalDriverConfigurationProviderImpl(
```
`private`?

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeLocalDriverConfigurationProviderImpl.java:
```diff
@@ -0,0 +1,72 @@
+ * MultiRuntime configuration provider for Local/Local runtime.
+ * This is a test provider that wraps local runtime with multi runtime.
+ */
+public class MultiRuntimeLocalDriverConfigurationProviderImpl extends AbstractDriverConfigurationProvider {
```
Why `public`?

shulmanb replied:
Yes as we need to bind this in the actual app.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeLocalDriverConfigurationProviderImpl.java:
```diff
@@ -0,0 +1,72 @@
+  private final Set<String> rackNames;
+
+  @Inject
+  MultiRuntimeLocalDriverConfigurationProviderImpl(@Parameter(MaxNumberOfEvaluators.class) final int maxEvaluators,
```
`private`?

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeYarnLocalDriverConfigurationProviderImpl.java:
```diff
@@ -0,0 +1,85 @@
+/**
+ * MultiRuntime configuration provider for Yarn/Local runtime.
+ */
+public class MultiRuntimeYarnLocalDriverConfigurationProviderImpl extends AbstractDriverConfigurationProvider {
```
Why `public`?

shulmanb replied:
Yes as we need to bind this in the actual app.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeYarnLocalDriverConfigurationProviderImpl.java:
```diff
@@ -0,0 +1,85 @@
+  private final Set<String> rackNames;
+
+  @Inject
+  public MultiRuntimeYarnLocalDriverConfigurationProviderImpl(
```
Why `public`?

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeYarnLocalDriverConfigurationProviderImpl.java:
```diff
@@ -0,0 +1,85 @@
+            .set(YarnDriverConfiguration.JOB_IDENTIFIER, jobId)
+            .set(YarnDriverConfiguration.CLIENT_REMOTE_IDENTIFIER, clientRemoteId)
+            .set(YarnDriverConfiguration.JVM_HEAP_SLACK, this.jvmSlack);
+    String serializedYarnConfig = serializeConfiguration(yarnModule);
```
please make this, and all other local variables, `final` wherever possible.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/MultiRuntimeResourceReleaseHandler.java:
```diff
@@ -0,0 +1,45 @@
+ */
+@Private
+@DriverSide
+public final class MultiRuntimeResourceReleaseHandler implements ResourceReleaseHandler {
```
Why `public`?

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,190 @@
+/**
+ * Hosts the actual runtime implementations and delegates invocations to them.
+ */
+public class RuntimesHost {
```
Why `public`? Also, make it `final`.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,190 @@
+ */
+public class RuntimesHost {
+  private final Set<String> runtimeDefinitions;
+  private Map<String, Injector> runtimesInjectors;
```
`final`?

shulmanb replied:
cannot be final as initialized on first invocation.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,190 @@
+  private final Set<String> runtimeDefinitions;
+  private Map<String, Injector> runtimesInjectors;
+  private final Injector originalInjector;
+  private String defaultRuntimeName;
```
`final`?

shulmanb replied:
Cannot be final as initialized on first invocation and not in c'tor

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,190 @@
+  private final Injector originalInjector;
+  private String defaultRuntimeName;
+
+  @Inject
```
`private`?

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,190 @@
+
+  public void onNext(final ResourceLaunchEvent value) {
+    try {
+      System.err.println("Resource launch called for runtime " + value.getRuntimeName());
```
Please replace all calls to `System.err` and `System.out` with logging.

shulmanb replied:
removed, don't really need logging there.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,190 @@
+        runtimeInjector.getInstance(ResourceManagerStartHandler.class).onNext(value);
+      }
+    } catch (InjectionException e) {
+      e.printStackTrace();
```
Please don't swallow exceptions, here and elsewhere. Also, we have gone through some work to make sure that exceptions in Java get logged exactly once.

Also, it would be better to catch the exception at each level.

shulmanb replied:
Added rethrowing RuntimeExceptions from handlers. I don't see a good reason to catch IOException in a private method and rethrow it as a RuntimeException, I think it makes method signature unclean as you don't know what to expect from it. And from other hand I don't see a reason to define my own exception to rethrow IOException, as I cannot throw it from handlers anyway.

markusweimer commented:
@shulmanb I've done a coding style-level review.

shulmanb commented:
@markusweimer done working on PR comments.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeHDILocalDriverConfigurationProviderImpl.java:
```diff
@@ -0,0 +1,83 @@
+/**
+ * MultiRuntime configuration provider for HDI/Local runtime.
+ */
+public final class MultiRuntimeHDILocalDriverConfigurationProviderImpl extends AbstractDriverConfigurationProvider {
```
Why are these, and the other `*Impl` classes `public`?

shulmanb replied:
They are public because we need to bind them. For example:
ConfigurationModule cm =  ExtensibleYarnClientConfiguration.getConfigurationModule(params)
            .set(ExtensibleYarnClientConfiguration.DRIVER_CONFIGURATION_PROVIDER,
                    MultiRuntimeYarnLocalDriverConfigurationProviderImpl.class);

In this case I need to bind this implementation to the YarnConfiguration. This way I am changing teh behavior of YarnSubmissionHandler.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,185 @@
+    this.runtimesInjectors = new HashMap<>();
+    final AvroConfigurationSerializer serializer = new AvroConfigurationSerializer();
+
+    for (String serializedRuntimeDefinition : runtimeDefinitions) {
```
`final`

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,185 @@
+
+    for (String serializedRuntimeDefinition : runtimeDefinitions) {
+
+      Configuration config = null;
```
Make these, and all other local variables, `final`, please.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,185 @@
+    return this.runtimesInjectors.get(runtimeName);
+  }
+
+  public void onNext(final ResourceLaunchEvent value) {
```
Do these `OnNext()` methods have to be `public`? Also, please document them. Further, it would help to rename them to `OnResourceLaunch(..)` etc. That way, browsing the code becomes easier. 

We've had similar code with far too many overloaded `OnNext()` calls in the past, and it became a bit of a review nightmare :-)

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,185 @@
+
+  public void onNext(final ResourceLaunchEvent value) {
+    try {
+      getInjector(value.getRuntimeName()).getInstance(ResourceLaunchHandler.class).onNext(value);
```
This uses Tang for each of the lookups, and needs two hash table lookups per event. Have you considered creating a runtime interface? E.g.:

```java
interface Runtime{
  String getRuntimeName();
  void onResourceLaunchEvent();
  ...
}
```

Then, you could have either a single implementation that gets passed in the various handlers at runtime or have one implementation of this interface per runtime.

shulmanb replied:
Agree, it will look much nicer. Two vs. one lookup doesn't matter in this case as this is control plane, even the fact that we will be able to cache handlers instead of doing tang lookup every time probably does not matter from perf point of view, but still looks much nicer.

markusweimer commented:
I've done another pass.

I have to think some more about the passing of hash maps in `public` interfaces. This is the second time we introduce a somewhat dirty `public` API as part of this new feature, which to me indicates that we are doing something not quite right here.

shulmanb commented:
@markusweimer Another round of fixes is in

afchung commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeConfigurationBuilder.java:
```diff
@@ -0,0 +1,101 @@
+    return this;
+  }
+
+  public MultiRuntimeConfigurationBuilder addDefaultRuntime(final String runtimeName) {
```
`setDefaultRuntime`.

afchung commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeConfigurationBuilder.java:
```diff
@@ -0,0 +1,101 @@
+  private static final Set<String> SUPPORTED_DEFAULT_RUNTIMES = new TreeSet<>(Arrays.asList(
+          org.apache.reef.runtime.yarn.driver.RuntimeIdentifier.RUNTIME_NAME));
+
+  private Set<String> runtimeNames = new TreeSet<>();
```
Why not `HashSet`?

shulmanb replied:
Changed.

afchung commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeConfigurationBuilder.java:
```diff
@@ -0,0 +1,101 @@
+ * A builder for Multi Runtime Configuration.
+ */
+public final class MultiRuntimeConfigurationBuilder {
+  private static final Set<String> SUPPORTED_RUNTIMES = new TreeSet<>(Arrays.asList(
```
Is this extensible beyond the hybrid of a `YARN` and `Local` runtime?

shulmanb replied:
Yes it is. To support that we will need to extend the builder and implement proper configuration module for additional runtimes. I am testing only yarn/local and local only variations.

afchung commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeConfigurationBuilder.java:
```diff
@@ -0,0 +1,101 @@
+    this.namedParameters.put(namedParameter, namedParameterValue);
+  }
+
+  public MultiRuntimeConfigurationBuilder addRuntime(final String runtimeName) {
```
Why not just use `enum` such that the supported types are obvious?

shulmanb replied:
Where this enum would be defined? I prefer to have that defined for each runtime.

afchung replied:
In your multi runtime module.

shulmanb replied:
That won't work because I need the actual runtime (local, yarn etc) to use the same runtime name when it creates evaluators. That's why I prefered having a convention of definiting RUNTIME_IDENTIFIER constant for every runtime.

markusweimer replied:
Why not `addRuntime(String id, Configuration runtimeConfiguration)`? That way, it would be extensible from the get go. Also, we wouldn't need the APIs specific to the local runtime in this class.

markusweimer replied:
Also, please document all `public` members.

afchung commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeConfigurationBuilder.java:
```diff
@@ -0,0 +1,101 @@
+    this.runtimeNames.remove(this.defaultRuntime);
+
+    // Currently only local runtime is supported as a secondary runtime
+    ConfigurationModule cm = ExtensibleYarnClientConfiguration.getConfigurationModule(this.namedParameters)
```
Why not just have a `Configuration` in the builder and merge the `Configurations` after `cm.build()`? Can we rename `cm`?

shulmanb replied:
Not sure what do you mean by that. I have a config module that I need to bind named parameters into, merging configs won't work here.

afchung replied:
Why not? You can define a config module in this class, bind your named parameters here, call `cm.build()`, and then merge the two built configurations.

shulmanb replied:
This named parameters are for other config module so I cannot define them here. I need to bind them on behalf of different module. So only thing I can do is pass them in and bind them in different place. Had long f2f discussion with Markus about this and couldn't find any other solution.

afchung replied:
It doesn't matter where the named parameter is bound, as long as the final Configuration contains some ConfigurationModule that has it bound, it will be fine.

shulmanb replied:
The problem is that I cannot merge these configurations. I am passing the actual runtime configs in a serialized way and I can instantiate them only in separate injectors as they define same parameters (with different values). So I need to bind these parameters in the original module and serialize it so I could instantiate it later in different injector (even not forked from the original). That's why this approach does not work. I tried it first.

afchung commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeConfigurationBuilder.java:
```diff
@@ -0,0 +1,101 @@
+            .set(ExtensibleYarnClientConfiguration.DRIVER_CONFIGURATION_PROVIDER,
+                    MultiRuntimeYarnLocalDriverConfigurationProviderImpl.class);
+
+    return cm.build();
```
Use chaining.

shulmanb replied:
done.

afchung commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/AbstractDriverConfigurationProvider.java:
```diff
@@ -0,0 +1,117 @@
+ */
+abstract class AbstractDriverConfigurationProvider implements DriverConfigurationProvider {
+
+  protected static String serializeConfiguration(final ConfigurationModule configModule) {
```
I don't think this belongs in this class. It's a simple function that generalizes across different usages.

shulmanb replied:
I don't agree with this statement :).This is an abstract class that all multi runtime configurations derive from.

afchung commented on lang/java/reef-runtime-multi/src/test/java/org/apache/reef/runtime/multi/driver/RuntimesHostTest.java:
```diff
@@ -0,0 +1,370 @@
+            new SpecificDatumWriter<>(RuntimeDefinition.class);
+    final String serializedConfiguration;
+    try (final ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+      final JsonEncoder encoder = EncoderFactory.get().jsonEncoder(RuntimeDefinition.SCHEMA$, out);
```
`rd.getSchema()`.

afchung commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/AbstractDriverConfigurationProvider.java:
```diff
@@ -0,0 +1,117 @@
+            new SpecificDatumWriter<>(RuntimeDefinition.class);
+    final String serializedConfiguration;
+    try (final ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+      final JsonEncoder encoder = EncoderFactory.get().jsonEncoder(RuntimeDefinition.SCHEMA$, out);
```
`rd.getSchema()`.

afchung commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/AbstractDriverConfigurationProvider.java:
```diff
@@ -0,0 +1,117 @@
+      configurationWriter.write(rd, encoder);
+      encoder.flush();
+      out.flush();
+      serializedConfiguration = out.toString("ISO-8859-1");
```
Do we have a good place to put this `String`?

shulmanb replied:
Not sure. Will define a const for that class.

shulmanb commented:
@afchung Uploaded fixes for your comments (and explanations for ones I didn't fix)

shulmanb commented:
@markusweimer I removed the delayed task and added a builder, please take a look.

afchung commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,169 @@
+  }
+
+  private synchronized void initialize() {
+    if (this.runtimes != null) {
```
What is this check for? Will `onRuntimeStart` be called more than once?

shulmanb replied:
Only if onStart will be called twice. not supposed to be called more then once, but I prefer having this check.

afchung commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,169 @@
+      try {
+        rd = parseSerializedRuntimeDefinition(serializedRuntimeDefinition);
+
+        if (rd.getDefaultConfiguration()) {
```
`isDefaultConfiguration()`?

shulmanb replied:
This is avro generated code.

afchung commented on lang/java/reef-runtime-multi/src/main/avro/RuntimeDefinition.avsc:
```diff
@@ -0,0 +1,34 @@
+    "fields":[
+      {"name":"runtimeName","type":"string"},
+      {"name":"serializedConfiguration","type":"string"},
+      {"name":"defaultConfiguration","type":"boolean"}
```
What does this do? Can you add some documentation by using the `"doc"` attribute?

shulmanb replied:
Done

markusweimer commented:
Will do. I might have a fix for [REEF-1229](https://issues.apache.org/jira/browse/REEF-1229) within the hour, in which case that should probably go in first such that tests of this can pass.

afchung commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,169 @@
+
+      final Injector rootInjector = Tang.Factory.getTang().newInjector();
+      try {
+        initializeInjector(rootInjector);
```
Why are we creating a new injector here?

shulmanb replied:
We need new injector for each runtime as they define same named parameters which conflict with each other, so we need to isolate them in different injectors.

afchung commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,169 @@
+      final Injector rootInjector = Tang.Factory.getTang().newInjector();
+      try {
+        initializeInjector(rootInjector);
+        final JavaConfigurationBuilder cb = Tang.Factory.getTang().newConfigurationBuilder();
```
Can we use a `ConfigurationModule` here instead?

shulmanb replied:
Don't think so.

afchung commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,169 @@
+            runtimeStatusEventHandler);
+  }
+
+  private RuntimeDefinition parseSerializedRuntimeDefinition(final String serializedRuntimeDefinition) throws
```
`static`

markusweimer commented:
I filled #866 for the PID store bug.

markusweimer commented:
@shulmanb Is this ready for review again?

shulmanb commented:
@markusweimer Yes. It is ready for review.

markusweimer commented:
@shulmanb thanks, I will do a pass now.

@afchung Can you do one after me?

markusweimer commented on lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloREEFMultiYarn.java:
```diff
@@ -0,0 +1,90 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.reef.examples.hello;
```
Can you move these new classes into their own package, please? the `.hello` namespace is the first spot people look at when learning REEF, and these classes aren't exactly first day reading :)

markusweimer commented on lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloREEFMultiYarn.java:
```diff
@@ -0,0 +1,90 @@
+  }
+
+  private static Configuration getHybridYarnSubmissionRuntimeConfiguration() {
+    MultiRuntimeConfigurationBuilder mrcb = new MultiRuntimeConfigurationBuilder();
```
Either make this `final` or make it one chained call in the `return` statement.

markusweimer commented on lang/java/reef-runtime-multi/src/main/avro/RuntimeDefinition.avsc:
```diff
@@ -0,0 +1,47 @@
+        "doc":"Serialized Tang configuration"
+      },
+      {
+        "name":"defaultConfiguration",
```
How do we ensure that only one of them is marked as default?

shulmanb replied:
1) Public builder interface exposes setter that throws if invoked more then once
2) RuntimesHost will throw if more then on default runtime was defined (added that)

markusweimer replied:
Yes, but that requires careful code editing in the future. Have you considered expressing this in the schema? E.g. by the approach I suggested earlier: Have a second schema for a collection of these `RuntimeDefinition` instances. That way, you can enforce naming only one as the default there, and won't have to repeat yourself in the code.

markusweimer commented on lang/java/reef-runtime-multi/src/main/avro/RuntimeDefinition.avsc:
```diff
@@ -0,0 +1,47 @@
+ * to the runtimes host
+ */
+ {
+    "namespace":"org.apache.reef.runtime.multi.utils",
```
Howe about `schemas` or `avro` as the package? That way, this code and the other code wouldn't be intermingled?

shulmanb replied:
Will add .avro to the package name

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/AbstractDriverConfigurationProvider.java:
```diff
@@ -0,0 +1,119 @@
+  protected static String serializeRuntimeDefinition(final String serializedConfig,
+                                                     final boolean defaultRuntime,
+                                                     final String runtimeName) {
+    RuntimeDefinition rd =
```
`final`. IntelliJ has an inspector that can find all cases of variables that can be, but are not, `final`. Please apply that to this change.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/AbstractDriverConfigurationProvider.java:
```diff
@@ -0,0 +1,119 @@
+ */
+abstract class AbstractDriverConfigurationProvider implements DriverConfigurationProvider {
+
+  private static final String CHARSET_NAME = "ISO-8859-1";
```
Why not `UTF-8`?

shulmanb replied:
This is what is used in AvroClassHierarchySerializer. Thought its a good idea to use same charset.

markusweimer replied:
Hmm, that would be an issue in its own right. Whenever possible, UTF-8 seems like the right choice to me.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/AbstractDriverConfigurationProvider.java:
```diff
@@ -0,0 +1,119 @@
+    return serializer.toString(localDriverConfiguration);
+  }
+
+  protected static String serializeRuntimeDefinition(final String serializedConfig,
```
Wouldn't it be cleaner to define a class like `AvroConfigurationSerializer` for this?

shulmanb replied:
Agree

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/AbstractDriverConfigurationProvider.java:
```diff
@@ -0,0 +1,119 @@
+   * @param jobId          the job id
+   * @return array of configuration modules for runtimes, when the first element contains the default runtime
+   */
+  protected abstract String[] getDriverConfiguration(final URI jobFolder,
```
Why `String[]` as the return value and not `Collection<Configuration>`? Or `Iterable<Configuration>`? The serialization into strings seems like a implementation detail which currently leaks.

shulmanb replied:
The main reason is that it is up to implementation to decide what is the default runtime. I could make the API more complex and pass out a more complex structure or let the implementation to create RuntimeDefinition. I will change the interface to return RuntimeDefinition[]

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/AbstractDriverConfigurationProvider.java:
```diff
@@ -0,0 +1,119 @@
+/**
+ * Provides base class for driver configuration providers for multi runtimes.
+ */
+abstract class AbstractDriverConfigurationProvider implements DriverConfigurationProvider {
```
Can this be solved via composition instead of an `abstract` base class? Composition is preferable because of the [Fragile Base Class](https://en.wikipedia.org/wiki/Fragile_base_class) problem.

shulmanb replied:
I don't see how. It will result in code duplication between different implementations as the public method will be similar. I think abstract base class works better in this case, as it defines common behavior for all implementations. Composition in this case will result in code duplication or different implementation of the same logic.

markusweimer replied:
Can't the code duplication be avoided? I see two ways:

  1. Have only one `DriverConfigurationProvider` which gets an object injected that supplies the per-runtime code.
  2. Alternatively, provide a class that supplies the common functionality to all the different implementations of `DriverConfigurationProvider`.

Those are the common rewrites to avoid the fragile base class problem. Why can't they be applied here?

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/AbstractDriverConfigurationProvider.java:
```diff
@@ -0,0 +1,119 @@
+    ConfigurationModule module = MultiRuntimeDriverConfiguration.CONF
+            .set(MultiRuntimeDriverConfiguration.JOB_IDENTIFIER, jobId)
+            .set(MultiRuntimeDriverConfiguration.CLIENT_REMOTE_IDENTIFIER, clientRemoteId);
+    for (String config : serializedConfigurations) {
```
`final String`

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeConfigurationBuilder.java:
```diff
@@ -0,0 +1,94 @@
+
+  private Set<String> runtimeNames = new HashSet<>();
+  private String defaultRuntime = null;
+  private final HashMap<Class, Object> namedParameters = new HashMap<>();
```
Is this still needed? I hoped we'd escape the HashMap hell :)

shulmanb replied:
No we didn't. We agreed on adding a builder till the composition of configuration modules will be solved.

shulmanb replied:
This is fixed

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeConfigurationBuilder.java:
```diff
@@ -0,0 +1,94 @@
+
+  public MultiRuntimeConfigurationBuilder addRuntime(final String runtimeName) {
+    if (!SUPPORTED_RUNTIMES.contains(runtimeName)) {
+      throw new IllegalArgumentException("Unsupported Runtime " + runtimeName);
```
[Commons lang](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/Validate.html) provides a nice API to capture this sort of validation. We use it elsewhere in REEF and it is hence safe to use here.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeConfigurationBuilder.java:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  public Configuration build() {
+    if (defaultRuntime == null) {
```
You can use `Optional<T>` to capture variables like this.

shulmanb replied:
What? local evaluators? I don't want to define parameter for this.

markusweimer replied:
`Optional<T>` is not a tang construct. It expresses a variable that maybe present or not. Java 8 has it, and we have our own, given that we are on Java 7 right now.

shulmanb replied:
Got it.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeHDILocalDriverConfigurationProviderImpl.java:
```diff
@@ -0,0 +1,83 @@
+            serializedLocalConfig,
+            false,
+            org.apache.reef.runtime.local.driver.RuntimeIdentifier.RUNTIME_NAME);
+    return new String[]{serializedHDIConfiguration, serializedLocalConfiguration};
```
I have a broader design question about this: Does this mean that multi-runtime only supports combinations of the local runtime with one other runtime? That falls a bit short of the potential of this change, IMHO.

shulmanb replied:
No it is not. This is just what I implemented right now. Nothing prevents defining Yarn/Mesos for example.

markusweimer replied:
I don't follow. This code to me suggests that we need to change REEF whenever we want to support another runtime combo, right?

shulmanb replied:
Actually you are right. Redesigned it. Now for each runtime that we want to support we need to implement a hook that will allow it to be injected to multi-runtime, and for runtime that we want to support as a submission runtime we need  to implement another hook. Once these hooks are implemented we can combine the rutnimes in any way.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,169 @@
+    this.originalInjector = injector;
+  }
+
+  private static RuntimeDefinition parseSerializedRuntimeDefinition(final String serializedRuntimeDefinition) throws
```
This should be merged with its serialization counterpart into a class, preferably injected. That would also allow for unit tests of the (de-)serialization routines.

shulmanb replied:
Done.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,169 @@
+        rd = parseSerializedRuntimeDefinition(serializedRuntimeDefinition);
+
+        if (rd.getDefaultConfiguration()) {
+          this.defaultRuntimeName = rd.getRuntimeName().toString();
```
This needs a check whether this has been set already.

shulmanb replied:
Done.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,169 @@
+        final Injector runtimeInjector = rootInjector.forkInjector(config, cb.build());
+        this.runtimes.put(rd.getRuntimeName().toString(), runtimeInjector.getInstance(Runtime.class));
+      } catch (InjectionException e) {
+        throw new RuntimeException("Unable to initialize runtimes.", e);
```
Typo in the log message.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,169 @@
+    this.runtimes = new HashMap<>();
+    final AvroConfigurationSerializer serializer = new AvroConfigurationSerializer();
+
+    for (final String serializedRuntimeDefinition : runtimeDefinitions) {
```
Wouldn't it be easier to have a schema for the group of runtimes in addition to each runtime on its own? That way, you can add the `Configuration` instances directly and can also take care of the unique default runtime at the schema level instead of requiring code to check for it.

shulmanb replied:
What do you mean by adding Configuration instances directly? I will still need to validate that default runtime actually exists, right? So I don't see a lot of advantage here.

markusweimer replied:
The advantage is that we get one composite object in which we can encode all the invariants. Also, the object as a whole would be serializable, avoiding all the `for` loops around strings.

shulmanb replied:
Done

shulmanb replied:
Done

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,169 @@
+
+  private Runtime getRuntime(final String requestedRuntimeName) {
+    String runtimeName = requestedRuntimeName;
+    if (runtimeName.isEmpty()) {
```
What if `runtimeName` is empty? Commons lang has a `IsNullOrEmpty` equivalent.

shulmanb replied:
What do you mean? it cannot be null, will add this check regardless.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,169 @@
+    }
+  }
+
+  private void initializeInjector(final Injector runtimeInjector) throws InjectionException {
```
What exactly is this method supposed to do? I'm confused because a `runtimeInjector` in my inderstanding would already have bindings for all of the below. Hence, this would throw lots of exceptions. Clearly, I am missing something here.

shulmanb replied:
I need to create different injector for each runtime as they define conflicting bindings. Also I cannot fork the original injector because of the same reason. What I do I create new injectors and copy form the original injector what I need.

markusweimer replied:
Add that as documentation, please.

shulmanb replied:
Done

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/org/apache/reef/runtime/multi/driver/parameters/package-info.java:
```diff
@@ -0,0 +1,22 @@
+ * under the License.
+ */
+/**
+ * A local implementation of REEF that uses local JVMs for execution.
```
Update. This is the `parameters` package :)

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/package-info.java:
```diff
@@ -0,0 +1,22 @@
+ * under the License.
+ */
+/**
+ * A local implementation of REEF that uses local JVMs for execution.
```
Update :)

markusweimer commented:
Done with my pass.

@afchung: it is all yours :)

shulmanb commented:
@markusweimer @afchung Done with another round of comments fixes

jwang98052 commented on lang/java/reef-examples/src/main/java/org/apache/reef/examples/hellomultiruntime/HelloMultiDriver.java:
```diff
@@ -0,0 +1,96 @@
+ * The Driver code for the Hello REEF Application with Multi Runtime.
+ */
+@Unit
+public final class HelloMultiDriver {
```
How about HelloRuntimeDriver? From the name, my impression is we are creating multiple drivers.

shulmanb commented:
@afchung @markusweimer I removed that Map from interface following Andrew's suggestion.

markusweimer commented:
Done with another pass over my prior comments.

shulmanb commented:
@markusweimer I refactored the abstract class into composition and made runtime configurations composable so we won't need to implement a class for each combination of runtimes. Also fixed other comments. Please take a look, I think I got all the comments.

markusweimer commented on lang/java/reef-runtime-multi/src/main/avro/RuntimeDefinition.avsc:
```diff
@@ -0,0 +1,65 @@
+      },
+      {
+        "name":"serializedConfiguration",
+        "type":"string",
```
You could import the schema for `Configuration` and reference it here. That would eliminate the code to (de-)serialize the strings from the multi-runtime code.

@afchung You have been looking at separating the runtime definition from a job submission. This schema could be helpful for that as well, right? If so, we should consider "promoting" it to reef-common.

shulmanb replied:
@markusweimer Do you think we can create a JIRA for this change (including importing schema) and for now leave it as is?

markusweimer replied:
Yeah, this shouldn't prevent the PR from being merged. But passing strings around in schemas seems "dirty" to me, as the strings are serialized Avro objects themselves.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeConfigurationBuilder.java:
```diff
@@ -0,0 +1,151 @@
+/**
+ * A builder for Multi Runtime Configuration.
+ */
+public final class MultiRuntimeConfigurationBuilder {
```
This seems to be restricted to the combination of the local runtime with either the YARN or HDI runtime, right?

If so: What would it take to make this generic?

shulmanb replied:
I don't think it is. Once we will have all runtimes supported as a submission runtime we won't have any restrictions. NOthing prevents having Yarn/HDI runtime for example and I can also enable local as a primary but it doesn't make a lot of sense. In order to enable Mesos we will need to define ExtensibleConfiguration for it and test so it can be done once someone will be interested in it and will be able to test it. We can have a JIRA for enabling MEsos support for multi-runtime.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeConfigurationBuilder.java:
```diff
@@ -0,0 +1,151 @@
+   * @return The builder instance
+   */
+  public MultiRuntimeConfigurationBuilder setMaxEvaluatorsNumberForLocalRuntime(final int maxLocalEvaluators) {
+    Validate.isTrue(maxLocalEvaluators > 0, "Max evaluators number shoudl be greater then 0");
```
typo: `shoudl` -> `should`

markusweimer replied:
also `then` -> `than`

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeConfigurationBuilder.java:
```diff
@@ -0,0 +1,151 @@
+
+  private Set<String> runtimeNames = new HashSet<>();
+  private Optional<String> defaultRuntime = null;
+  private String submissionRunitme;
```
typo

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeConfigurationBuilder.java:
```diff
@@ -0,0 +1,151 @@
+
+  private final HashMap<Class, Object> namedParameters = new HashMap<>();
+
+  private Set<String> runtimeNames = new HashSet<>();
```
`final`

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeConfigurationBuilder.java:
```diff
@@ -0,0 +1,151 @@
+  private final HashMap<Class, Object> namedParameters = new HashMap<>();
+
+  private Set<String> runtimeNames = new HashSet<>();
+  private Optional<String> defaultRuntime = null;
```
Use `Optional.empty()` instead. Then, you can replace `null` checks with calls to `orElse()` below.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeDefinitionBuilder.java:
```diff
@@ -0,0 +1,91 @@
+/**
+ * Builder for multi runtime definition.
+ */
+public final class MultiRuntimeDefinitionBuilder {
```
This API is defined in terms of `ConfigurationModule`. I would have expected it to be defined in terms of `Configuration`. What's the reason to require the (not finalized) `ConfigurationModule` for a runtime to be passed in? As far as I can tell, the code immediately moves calls `.build()` on the `ConfigurationModule` anyhow.

shulmanb replied:
No reason for that, just didn't know that this is the right thing  to do. Fixed.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeDefinitionBuilder.java:
```diff
@@ -0,0 +1,91 @@
+  public MultiRuntimeDefinitionBuilder addRuntime(final ConfigurationModule config, final String runtimeName){
+    Validate.notNull(config, "runtime configuration module should not be null");
+    Validate.isTrue(!StringUtils.isEmpty(runtimeName) && !StringUtils.isBlank(runtimeName));
+    RuntimeDefinition rd = createRuntimeDefinition(config, runtimeName);
```
`final`

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeDefinitionBuilder.java:
```diff
@@ -0,0 +1,91 @@
+   * @return The builder instance
+   */
+  public MultiRuntimeDefinitionBuilder setDefaultRuntimeName(final String runtimeName){
+    Validate.isTrue(!StringUtils.isEmpty(runtimeName) && !StringUtils.isBlank(runtimeName), "runtimeName " +
```
This can be rewritten with [`StringUtils.isNotBlank(...)`](http://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/StringUtils.html#isNotBlank%28java.lang.String%29), right?

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeDefinitionBuilder.java:
```diff
@@ -0,0 +1,91 @@
+   */
+  public MultiRuntimeDefinition build(){
+    Validate.isTrue(this.runtimes.size() == 1 || !StringUtils.isEmpty(this.defaultRuntime), "Default runtime " +
+            "should be set if more then single runtime provided");
```
`then` -> `than a`

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/MultiRuntimeDefinitionBuilder.java:
```diff
@@ -0,0 +1,91 @@
+      this.defaultRuntime = this.runtimes.keySet().iterator().next();
+    }
+
+    Validate.isTrue(this.runtimes.containsKey(this.defaultRuntime), "Default runtime shoudl be configured");
```
`shoudl` -> `should`

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,171 @@
+  private final Injector originalInjector;
+  private String defaultRuntimeName;
+  private final String serializedRuntimeDefinition;
+  private MultiRuntimeDefinitionSerializer  runtimeDefinitionSerializer = new MultiRuntimeDefinitionSerializer();
```
`final`

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,171 @@
+final class RuntimesHost {
+  private Map<String, Runtime> runtimes;
+  private final Injector originalInjector;
+  private String defaultRuntimeName;
```
This should be possible to be `final`, right? At the time this object is created, we can know the value, I believe?

shulmanb replied:
Fixed for everything but runtimes which I prefer to initialize later

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,171 @@
+ * Hosts the actual runtime implementations and delegates invocations to them.
+ */
+final class RuntimesHost {
+  private Map<String, Runtime> runtimes;
```
This should be possible to be `final`, right? At the time this object is created, we can know the value, I believe?

shulmanb replied:
Actually no, I check for null in order to prevent multiple invocations of initialize method. I can define additional boolean for it, but don't see why I should do it.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,171 @@
+  /**
+   * Initializes teh configured runtimes.
+   */
+  private synchronized void initialize() {
```
Why not move all of this into the constructor?

shulmanb replied:
because I receive the injector in the constructor and Tang throws if I access it from the constructor.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,171 @@
+   */
+  private Runtime getRuntime(final String requestedRuntimeName) {
+    String runtimeName = requestedRuntimeName;
+    if (StringUtils.isEmpty(runtimeName) || StringUtils.isBlank(runtimeName)) {
```
[`isBlank`](http://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/StringUtils.html#isBlank%28java.lang.String%29) implies `isEmpty`. This can be rewritten as

```java
if (StringUtils.isBlank(runtimeName)) {
```

shulmanb replied:
right should have read better the docs... Will fix here and everywhere else.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java:
```diff
@@ -0,0 +1,171 @@
+      runtimeName = this.defaultRuntimeName;
+    }
+
+    return this.runtimes.get(runtimeName);
```
This trusts the value of `runtimeName`. It is probably safer to check for null here or to check at the callsite.

shulmanb replied:
Isn't that what I do here:
if (StringUtils.isEmpty(runtimeName) || StringUtils.isBlank(runtimeName))

markusweimer replied:
Not quite. If the string isn't blank (which is the only thing you need to check, as blank implies empty), it can still refer to a runtime that isn't in `this.runtimes`. In other words:

```java
    return this.runtimes.get(runtimeName);
```

could return `null` with no explanation why.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/utils/MultiRuntimeDefinitionSerializer.java:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Serializer for MultiRuntimeDefinition.
+ */
+public final class MultiRuntimeDefinitionSerializer {
```
Please add a test for this class.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/utils/MultiRuntimeDefinitionSerializer.java:
```diff
@@ -0,0 +1,73 @@
+   * @param runtimeDefinition the Avro object to serialize
+   * @return Serialized avro string
+   */
+  public String serialize(final MultiRuntimeDefinition runtimeDefinition){
```
in other cases, we have used `toString(...)` and `fromString(...)` for these method names. That allows us to add other serialization targets and removes ambiguity when reading the code at the call site.

shulmanb replied:
Sure. Will change.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/utils/MultiRuntimeDefinitionSerializer.java:
```diff
@@ -0,0 +1,73 @@
+   * @return Avro object
+   * @throws IOException
+   */
+  public MultiRuntimeDefinition deserialize(final String serializedRuntimeDefinition) throws
```
This declares the exception while the above converts it into a `RuntimeException`. It is probably good to harmonize those two cases. Personally, I would declare the exceptions as `IOExceptions` could be handled at the call site.

shulmanb replied:
I refactored to invoke this from the constructor, so I prefer to throw Unchecked exception. For the lower level I prefer to have IOException to have cleaner API there.

markusweimer commented on lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/utils/MultiRuntimeDefinitionSerializer.java:
```diff
@@ -0,0 +1,73 @@
+    final JsonDecoder decoder = DecoderFactory.get().
+            jsonDecoder(MultiRuntimeDefinition.getClassSchema(), serializedRuntimeDefinition);
+    final SpecificDatumReader<MultiRuntimeDefinition> reader = new SpecificDatumReader<>(MultiRuntimeDefinition.class);
+    MultiRuntimeDefinition rd = reader.read(null, decoder);
```
`final`

shulmanb replied:
Done.

markusweimer commented on lang/java/reef-runtime-multi/src/test/java/org/apache/reef/runtime/multi/driver/RuntimesHostTest.java:
```diff
@@ -0,0 +1,362 @@
+            SerializedRuntimeDefinition.class,
+            "corrupted avro");
+
+    Injector forked = injector.forkInjector(cb.build());
```
`final`

shulmanb replied:
Done.

markusweimer commented:
@shulmanb Thanks, I have done another pass and left some comments in the code.

More broadly speaking, my main concern is that the API *seems* to be restricted to combinations of the local runtime with YARN or HDI. Maybe I am wrong on that. If not: I don't see a hard reason for that restriction. If we can support any combo of runtimes, we should :)

On a similar note: This identifies a runtime with a `String` and that `String` is defined as a constant. This in turn means that each runtime can only be part of the multi runtime setup exactly once, right? Again, I don't think the design mandates that restriction. Can we lift it?

shulmanb commented:
@markusweimer Your first concern will be lifted once we will implement proper support for Mesos and other runtimes. I don't want to implement it right now (as I can't really test it) but we should create a Jira for it. IN general we will need to support all runtimes but I prefer to release it with partial support and add support in the future. WDYT?

I can see why we would want to support several yarn environmens but for now it won't really work right? As we depend on the local configuration. I don't see any reason to support several local environments for example. IN order to support it we will need to introduce user provided identifier for each runtime which is possible and pass this identifier to the actual runtime so it could be attached to the allocated evaluators. This is all doable but as before I would prefer to release it with current limitations (I don't really see any usecases for these two) and have a JIRA for them.

WDYT?

shulmanb commented:
@markusweimer Pushed fixes for minor comments and replied for the questions you have raised.

shulmanb commented:
@markusweimer do you see any additional blockers for this PR?

markusweimer commented:
@shulmanb, can you please check that the current state of the branch still works on YARN? You probably should merge or rebase to current master locally before running that test to make sure it is indicative of the run after we merged this into master :)

markusweimer commented:
Please *don't* push the merged / rebased version. It will mess up the substantial discussion on this thread.

shulmanb commented:
@markusweimer sure, will test now

shulmanb commented:
@markusweimer Yarn tests passed, HelloReefMultiYarn passed.

markusweimer commented:
I will test the .NET side and merge.

shulmanb commented:
@markusweimer Thanks.
